x-pack/plugins/security_solution/public/resolver/models/node_data.ts

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { FetchedNodeData, IDToNodeInfo, NodeData } from '../types';
+
+/**
+ * Creates a copy of the node data map and initializes the specified IDs to an empty object with status requested.
+ *
+ * @param storedNodeInfo the node data from state
+ * @param requestedNodes a set of IDs that are being requested
+ */
+export function setRequestedNodes(
+  storedNodeInfo: IDToNodeInfo | undefined,
+  requestedNodes: Set<string>
+): IDToNodeInfo {
+  const originalData = storedNodeInfo ?? [];
+  const requestedNodesArray = Array.from(requestedNodes);
+  return new Map<string, NodeData>([
+    ...originalData,
+    ...requestedNodesArray.map((id: string): [string, NodeData] => [
+      id,
+      { events: [], status: 'loading' },
+    ]),
+  ]);
+}
+
+/**
+ * Creates a copy of the node data map and sets the specified IDs to an error state.
+ *
+ * @param storedNodeInfo the node data from state
+ * @param errorNodes a set of IDs we requested from the backend that returned a failure
+ */
+export function setErrorNodes(
+  storedNodeInfo: IDToNodeInfo | undefined,
+  errorNodes: Set<string>
+): IDToNodeInfo {
+  const originalData = storedNodeInfo ?? [];
+  const errorNodesArray = Array.from(errorNodes);
+  return new Map<string, NodeData>([
+    ...originalData,
+    ...errorNodesArray.map((id: string): [string, NodeData] => [
+      id,
+      { events: [], status: 'error' },
+    ]),
+  ]);
+}
+
+/**
+ * Marks the node id to be reloaded by the middleware. It removes the entry in the map to mark it to be reloaded.
+ *
+ * @param storedNodeInfo the node data from state
+ * @param nodeID the ID to remove from state to mark it to be reloaded in the middleware.
+ */
+export function setReloadedNodes(
+  storedNodeInfo: IDToNodeInfo | undefined,
```

Comment:
I see a repeated pattern here of accepting `undefined` for `storedNodeInfo`. Perhaps you could use this instead:

```suggestion
  storedNodeInfo: IDToNodeInfo = [],
```